src/pycapsule.rs

```diff
@@ -0,0 +1,276 @@
+        destructor: Option<ffi::PyCapsule_Destructor>,
+    ) -> PyResult<&'py Self> {
+        let val = Box::new(value);
+
+        let cap_ptr = unsafe {
+            ffi::PyCapsule_New(Box::into_raw(val) as *mut c_void, name.as_ptr(), destructor)
+        };
+        if cap_ptr.is_null() {
+            Err(PyErr::fetch(py))
+        } else {
+            Ok(unsafe { py.from_owned_ptr::<PyCapsule>(cap_ptr) })
+        }
+    }
+
+    /// Import an existing capsule.
+    ///
+    /// The `name` should match the path to module attribute exactly in the form
+    /// of `module.attribute`, which should be the same as the name within the
+    /// capsule. `no_block` indicates to use
+    /// [PyImport_ImportModuleNoBlock()](https://docs.python.org/3/c-api/import.html#c.PyImport_ImportModuleNoBlock)
+    /// or [PyImport_ImportModule()](https://docs.python.org/3/c-api/import.html#c.PyImport_ImportModule)
+    /// when accessing the capsule.
+    pub fn import<'py, T>(py: Python<'py>, name: &CStr, no_block: bool) -> PyResult<&'py T> {
+        let ptr = unsafe { ffi::PyCapsule_Import(name.as_ptr(), no_block as c_int) };
+        if ptr.is_null() {
+            Err(PyErr::fetch(py))
+        } else {
+            Ok(unsafe { &*(ptr as *const T) })
+        }
+    }
+
+    /// Set a context pointer in the capsule to `T`
+    pub fn set_context<T>(&self, py: Python, context: T) -> PyResult<()> {
+        let ctx = Box::new(context);
+        let result =
+            unsafe { ffi::PyCapsule_SetContext(self.as_ptr(), Box::into_raw(ctx) as _) as u8 };
+        if result != 0 {
+            Err(PyErr::fetch(py))
+        } else {
+            Ok(())
+        }
+    }
+
+    /// Get a reference to the context `T` in the capsule, if any.
+    pub fn get_context<T>(&self, py: Python) -> PyResult<Option<&T>> {
+        let ctx = unsafe { ffi::PyCapsule_GetContext(self.as_ptr()) };
+        if ctx.is_null() {
+            if self.is_valid() & PyErr::occurred(py) {
+                Err(PyErr::fetch(py))
+            } else {
+                Ok(None)
+            }
+        } else {
+            Ok(Some(unsafe { &*(ctx as *const T) }))
+        }
+    }
+
+    /// Obtain a reference to the value `T` of this capsule.
+    ///
+    /// # Safety
+    /// This is unsafe because there is no guarantee the pointer is `T`
+    pub unsafe fn reference<T>(&self) -> &T {
+        &*(self.get_pointer() as *const T)
+    }
+
+    /// Get the raw `c_void` pointer to the value in this capsule.
+    pub fn get_pointer(&self) -> *mut c_void {
+        unsafe { ffi::PyCapsule_GetPointer(self.0.as_ptr(), self.name().as_ptr()) }
+    }
+
+    /// Check if this is a valid capsule.
+    pub fn is_valid(&self) -> bool {
+        let r = unsafe { ffi::PyCapsule_IsValid(self.as_ptr(), self.name().as_ptr()) } as u8;
+        r != 0
+    }
+
+    /// Get the capsule deconstructor, if any.
+    pub fn get_deconstructor(&self, py: Python) -> PyResult<Option<ffi::PyCapsule_Destructor>> {
```

Comment:
Typo: `deconstructor` -> `destructor`.

```suggestion
    /// Get the capsule destructor, if any.
    pub fn get_destructor(&self, py: Python) -> PyResult<Option<ffi::PyCapsule_Destructor>> {
```